lib/Dialect/TTNN/Analysis/ShardSolver.cpp

```diff
@@ -561,9 +566,104 @@ bool ShardSolver::checkShardCompatible(
+// usage (Layout0GridVolume, Layout1GridVolume, Layout2GridVolume).
+//
+//    Op0 ----- (4, 8, 2)
+//     |
+//    Op1 ----- (8, 4, 2)
+//    / \
+//   /   \
+//  Op2  Op3 -- (4, 4, 2) (4, 4, 2)
+//   \   /
+//    \ /
+//    Op4 ----- (2, 1, 1)
+//     |
+//    Op5 ----- (2, 1, 1)
+//
+// Here is how structure looks after preprocessing is complete:
+//
+//    Op0 ----- (24, 22, 10)
+//     |
+//    Op1 ----- (20, 14, 8)
+//    / \
+//   /   \
+//  Op2  Op3 -- (6, 5, 3) (6, 5, 3)
+//   \   /
+//    \ /
+//    Op4 ----- (4, 2, 2)
+//     |
+//    Op5 ----- (2, 1, 1)
+//
+// Global max of 24 core usage is achieved by selecting layout[0] for each Op.
+//
+// Returns map of op to vector of max core usage for each layout.
+llvm::DenseMap<Operation *, SmallVector<float, 64>>
+ShardSolver::produceMaxCoreUsage() {
+  using Paths = llvm::SmallVector<Path, 16>;
+  llvm::DenseMap<Operation *, SmallVector<float, 64>> accCoreUsage(
+      shardedOps->size());
+
+  // Start from the tail of the chain and build up the max core usage(schedule
+  // in backwards).
+  //
+  for (auto shardSpec = shardSpecs->rbegin(); shardSpec != shardSpecs->rend();
+       ++shardSpec) {
+    Operation *op = shardSpec->op;
+    std::vector<tt::LayoutAttr> const &layouts = getLegalLayouts(op);
+    assert(!layouts.empty());
+
+    // Find the layout that leads to the max core usage.
+    // Start with grid volume of current op.
+    //
+    for (size_t i = 0; i < layouts.size(); ++i) {
+      tt::LayoutAttr const &layout = layouts[i];
+      uint64_t coreUsage = layout.getGrid().getGridVolume();
+      accCoreUsage[op].push_back(coreUsage);
+    }
+
+    // Add core usage of current op users via live path connections.
+    //
+    SmallVector<ShardSolver::PathSet *> userPathSets = getUserPathSetsPts(op);
+    for (size_t i = 0; i < userPathSets.size(); ++i) {
+      ShardSolver::PathSet *pathSet = userPathSets[i];
+      const Paths &paths = pathSet->getPaths();
+      SmallVector<uint64_t, 64> maxCoreUsage(layouts.size(), 0);
+      Operation *consumerOp = pathSet->getConsumerOp();
+      size_t consumerInChainOperandSize =
+          getOperandPathSetsPts(consumerOp).size();
+      uint64_t consumerCoreUsage = 0;
+      for (auto const &path : paths) {
+        assert(bitsets[bitsetIds[op]].test(path.producerId));
+        assert(bitsets[bitsetIds[consumerOp]].test(path.consumerId));
+        consumerCoreUsage = accCoreUsage[consumerOp][path.consumerId];
+        if (consumerCoreUsage > maxCoreUsage[path.producerId]) {
+          maxCoreUsage[path.producerId] = consumerCoreUsage;
+        }
+      }
+
+      for (size_t i = 0; i < layouts.size(); ++i) {
+        // Add max core usage of consumer ops to current op layout.
+        // We divide by consumerInChainOperandSize to normalize the core usage
+        // based on forking factor(so that cores are not counted more than
+        // once).
+        //
+        accCoreUsage[op][i] += static_cast<float>(maxCoreUsage[i]) /
```

Comment:
As discussed offline, this can produces incorrect results (have some layout core usage scaled) when the chain contains joins without forks. Seems totally fine for now, but let's just leave a comment as a reminder.

Reply:
Commented.